### PR TITLE
Enable Bazel build system support

### PR DESCRIPTION
**Pull request by WSUFan**
This pull request enables Bazel build support for cpptrace. Please note that currently, only Linux is supported. Additional platform support will be added if necessary.

To build with bazel, simply run
```bash
bazel build //...
```

To test with bazel, simply run
```bash
bazel test //...
```

### CODE REVIEW

File: .github/workflows/build.yml
Comment (jeremy-rifkin) [inline .github/workflows/build.yml:36]: Is it running the same command twice?

File: .github/workflows/test.yml
Comment (jeremy-rifkin) [inline .github/workflows/test.yml:160]: Same here

File: BUILD
Comment (jeremy-rifkin) [inline BUILD:8]: It's unfortunate to hard-code this, however it's pragmatic.

**Review by jeremy-rifkin — COMMENTED** [review]
Thank you for taking the time to put this together, I have a couple quick questions and then I think this looks good to me

File: .github/workflows/build.yml
Comment (WSUFan) [reply .github/workflows/build.yml:36]: Oh this is typo. I was thinking to build cpptrace in both debug and release mode. Or do you think only one of them is enough?

File: .github/workflows/test.yml
Comment (WSUFan) [reply .github/workflows/test.yml:160]: Yeah. Perform the unit test in both debug and release mode.

Comment (jeremy-rifkin) [reply .github/workflows/test.yml:160]: Oh my bad, this I misread (and I can't stop reading `//..` as a comment 😄)

File: .github/workflows/build.yml
Comment (jeremy-rifkin) [reply .github/workflows/build.yml:36]: Either way is fine with me, it doesn't hurt to do both

File: BUILD
Comment (WSUFan) [reply BUILD:8]: Yeah, this hardcoded and unfortunate I think this is the Bazel way. Currently only LIBDWARF is supported.... Or do you think we should support all backend in one shot?

**Comment by WSUFan** [comment]
Can you approve Github workflows for this PR? Let's see how it goes.
image

**Comment by jeremy-rifkin** [comment]
I can

**Comment by jeremy-rifkin** [comment]
Looks like there were a lot more test failures than I would have expected due to trying to build cmake in a folder called `build`. That's really unfortunate, I don't like case-insensitive filesystems. Calling it `BUILD.bazel` might be best?

**Comment by WSUFan** [comment]
> Looks like there were a lot more test failures than I would have expected due to trying to build cmake in a folder called `build`. That's really unfortunate, I don't like case-insensitive filesystems. Calling it `BUILD.bazel` might be best?

Yeah, I changed the filename. But to run Github workflows, still need another approval? :(

File: .github/workflows/build.yml
Comment (jeremy-rifkin) [inline .github/workflows/build.yml:33]: Why is ncurses needed?

File: BUILD.bazel
Comment (jeremy-rifkin) [inline BUILD.bazel:10]: Circling back to this, happy to go with this for now but it'd be nice to let users choose other back-ends if they want (e.g. libunwind). How hard would that be to support?

**Review by jeremy-rifkin — APPROVED** [review]
I think this looks good to me, two quick questions

File: .github/workflows/build.yml
Comment (WSUFan) [reply .github/workflows/build.yml]: Clang16 needs this. But this is not a problem anymore because I upgraded the clang to 18 which does not need this lib.

File: BUILD.bazel
Comment (WSUFan) [reply BUILD.bazel:10]: Letting the user choose is always the best approach. The most challenging part is converting other backends to be built with Bazel. Fortunately, Bazel's `foreign_cc` rule can help us. I'll try to make other backends work

Comment (jeremy-rifkin) [reply BUILD.bazel:10]: Thank you so much, I really appreciate all the work! If it proves nontrivial, libunwind is the main back-end that would be useful to support for users as it's the only way to do signal-safe unwinding.

Comment (WSUFan) [reply BUILD.bazel:10]: You're welcome! Could you also enable this PR to automatically trigger the GitHub workflows? That way, I can test it myself.

Comment (jeremy-rifkin) [reply BUILD.bazel:10]: I don't see a way to not require me to hit approve, but I can hit approve

File: .github/workflows/build.yml
Comment (WSUFan) [reply .github/workflows/build.yml]: It looks like clang-18 still depends on this lib... so we sill need to install.

**Review by jeremy-rifkin — APPROVED** [review]

**Comment by jeremy-rifkin** [comment]
It looks like CI is failing to build, happy to merge once it builds

**Comment by WSUFan** [comment]
> It looks like CI is failing to build, happy to merge once it builds

Yeah. This is strange because I tested it successfully in my env (using ubuntu 22.04 image). 
But I changed some build configs and let's see.

**Comment by jeremy-rifkin** [comment]
It looks like that passed, sweet! Thanks again for taking the time to set this up!

**Pull request merged**
